### PR TITLE
style: apply black

### PR DESCRIPTION
**Pull request by henryiii**
This enables black. String normalization is turned off until the last commit to keep the diffs down, so the focus can be on structure rather than a bunch of quote changes (seems to change docstrings anyway, though, FYI). We also have a longer than normal line length set, compared to last time this might have been attempted.

Having a diff-optimized and highly restricted style might be especially useful before we split for 2.x development. What do others think? The auto-fixes from pre-commit are especially nice with something like black over flake8.

### CODE REVIEW

File: bin/update_dependencies.py
Comment (henryiii) [inline bin/update_dependencies.py:92]: This is still readable without the extra lines, but if we wanted to keep it, we can add fmt on/off comments.

**Comment by joerick** [comment]
I'm not so interested in debating individual merits of the transform, debating style is a great way to waste programmers' time! So I'll keep my feedback high-level.

I'm -1 on this. I agree that Black has some nice properties regarding merges, but for me, the intelligibility of the code is more important, and I think that the Black'd code is harder to read+understand, and I would find it harder to work with.

So I'm in favour of closing this one. Would the other maintainers agree/disagree with the above - @YannickJadoul, @mayeut?

**Comment by YannickJadoul** [comment]
I had - for once - successfully stopped myself pointing out a few places where I think black is ... well, doing a worse job than we have. (I can still do so now, if you like?)
There's a few nice cleanups as well, though, that we could maybe keep (and maybe we'll move towards a more grey'ish kind of style anyway, over time?), but yeah, still not a huge fan of such a massive PR.

File: bin/bump_version.py
Comment (YannickJadoul) [inline bin/bump_version.py:48]: This is worse, IMO.

File: bin/bump_version.py
Comment (YannickJadoul) [inline bin/bump_version.py:62]: So, black ignores `# noqa`?

(I don't mind this change, btw; it does look funky, but someone seems to have intentionally written this? So I don't like any automatic thing overriding the intentional style decisions made by a human. (At least not until we have a few more AI breakthroughs.))

File: bin/bump_version.py
Comment (YannickJadoul) [inline bin/bump_version.py:89]: These kinds of things are a good catch, though.

File: bin/bump_version.py
Comment (YannickJadoul) [inline bin/bump_version.py:102]: Again, I don't see how this is better/cleaner than before.

This should be an f-string, anyway, but that's besides the point, I guess.

File: bin/bump_version.py
Comment (YannickJadoul) [inline bin/bump_version.py:124]: There seems to be a +- consistent style (I guess by @joerick?) where flags of a subprocess call are always on new lines. So this actually introduces inconsistency one some level.

File: bin/inspect_all_known_projects.py
Comment (YannickJadoul) [inline bin/inspect_all_known_projects.py:65]: Yes. But we probably have black (with a smaller maximum line length) to blame for the version we now have, already?

File: bin/make_dependency_update_pr.py
Comment (YannickJadoul) [inline bin/make_dependency_update_pr.py:64]: Again, the left side makes more sense to me, having `'--repo', 'joerick/cibuildwheel'` on one line. And it's kind of hard to explain to black?

File: bin/projects.py
Comment (YannickJadoul) [inline bin/projects.py:83]: There's a reason there's two lines, here: the actual string is also two lines long.

File: cibuildwheel/__main__.py
Comment (YannickJadoul) [inline cibuildwheel/__main__.py:424]: I'm happy having this on one line. All you need when scrolling through the code is there: you're appending a warning. If part of the line is off screen, eh, who cares unless you're in the rare case of editing the warning message?

File: cibuildwheel/typing.py
Comment (YannickJadoul) [inline cibuildwheel/typing.py:12]: This might not be a bad idea?

**Review by YannickJadoul — COMMENTED** [review]
A not-so-random sample

File: bin/bump_version.py
Comment (henryiii) [reply bin/bump_version.py:48]: Why? It's_ really_ easy to miss the `{}` in the original version. Black ensures you don't "helpfully" hide things like this. I'm strongly in favor of the rewritten version, actually, since if I'm reading quickly, I get that it's a set, while in the original version, I'd have to look very carefully.

File: bin/bump_version.py
Comment (henryiii) [reply bin/bump_version.py:62]: I think that it was totally a mistake? It doesn't line up with anything else...

And yes, black doesn't care about noqa, that's for flake8. It cares about style: on / style: off.

File: bin/bump_version.py
Comment (henryiii) [reply bin/bump_version.py:124]: If we did want to go with black, we could add fmt: on / off around these. It's okay to have special style for certain things and consistent style everywhere else. Or add a trailing comma, which would do one per line.

File: bin/inspect_all_known_projects.py
Comment (henryiii) [reply bin/inspect_all_known_projects.py:65]: The other version is fine, too.

File: bin/projects.py
Comment (henryiii) [reply bin/projects.py:83]: Python's merging of strings that follow each other is _highly_ prone to mistakes. I last caught one five days ago https://github.com/scikit-hep/decaylanguage/pull/127#discussion_r600857204

I think if you added a `+`, this would format correctly. Or you could add a fmt on/off to clarify you know what you are doing.

File: cibuildwheel/__main__.py
Comment (henryiii) [reply cibuildwheel/__main__.py:424]: this should be `msg = ...`, `warnings.append(msg)`, for consistency with raise statements. Then I think it would probably go on one line.

File: bin/bump_version.py
Comment (YannickJadoul) [reply bin/bump_version.py:48]: Is it so important that it's a dict? `'q': ` should also clarify that, though?

Comment (henryiii) [reply bin/bump_version.py:48]: I missed that it was a dict, I thought it was a set. So probably no? :P

File: bin/bump_version.py
Comment (YannickJadoul) [reply bin/bump_version.py:62]: The colon lines up with the colon of `'    New version: '`. Not saying I'm a fan, btw, but someone care enough to do this and add `# noqa`.

> It cares about style: on / style: off.

Hmmm, that's quite heavy :-/

File: bin/bump_version.py
Comment (YannickJadoul) [reply bin/bump_version.py:124]: > Or add a trailing comma, which would do one per line.

The thing is that things that go together (`'-m'` and `f'v{new_version}'`) are together here, though.

File: bin/inspect_all_known_projects.py
Comment (YannickJadoul) [reply bin/inspect_all_known_projects.py:65]: It's a bit long, though. If it's a simple one-step concept (i.e., get the contents of this file in this repo), it's nice if it fits on one line.

File: bin/projects.py
Comment (YannickJadoul) [reply bin/projects.py:83]: That's tricky though, yes. But there's still a difference with the issue you refer to.

> Or you could add a fmt on/off to clarify you know what you are doing.

I still don't like that, though. It distracts from what's actually going on, to me.

File: bin/bump_version.py
Comment (YannickJadoul) [reply bin/bump_version.py:48]: :-D Black doesn't help there, though, but I guess that's besides the point.

File: bin/inspect_all_known_projects.py
Comment (henryiii) [reply bin/inspect_all_known_projects.py:65]: Not if that line can't be read in your editor without scrolling. Editors don't widen out when you come to a longer line just because it's a single concept.

File: bin/bump_version.py
Comment (henryiii) [reply bin/bump_version.py:124]: Then use `f"--message=v{new_version}"`. All together now, and even clearly says what is being done instead of "m". :)

Comment (henryiii) [reply bin/bump_version.py:124]: This is generally the solution, by the way - you should not use single char flags in a script, you should write it out - there's no need to save characters in a script. And if you do, it's then in one line instead of two using the `=` syntax.

File: bin/make_dependency_update_pr.py
Comment (henryiii) [reply bin/make_dependency_update_pr.py:64]: `"--repo=joerick/cibuildwheel"`, done.

**Comment by henryiii** [comment]
I can pick out the best changes and just make those, but then we lose pre-commit.ci's auto fixes for style. I feel like most of the arguments against black are "80% of the changes are good, but 20% are worse, so that's totally unacceptable" (and really, it's more like 80% good, 15% easily fixable, and 5% actually worse).

> debating style is a great way to waste programmers' time

That's the point of Black. No choices, no arguments, one style everywhere. :) Also, consider the PR an "Issue with an example". ;)

**Comment by YannickJadoul** [comment]
> "80% of the changes are good, but 20% are worse, so that's totally unacceptable"

If it got to 80%, I'd be quite happy, but not sure there. Currently for me it's seems more like 20% good, 50% meh, 30% "yikes, really?" (but those are numbers I literally just pulled out of the air).

**Comment by YannickJadoul** [comment]
> 30% "yikes, really?"

Well, not per se "yikes, really?". Also "no, we thought about that, there's a reason to present it this way".

**Comment by henryiii** [comment]
No more out of the air then mine were. ;)

**Comment by Czaki** [comment]
I really like black. It saves many problems on other projects in which I participate. But I see its weakness. 

One thing which I remember from the time where I was a more active contributor to `cibuildwheel` there was a pain of commits which fixes small style problem and push on queue another job which delays all active PR. 

Adding some style checker which will grab most of the problems with formating and block running half-hour tests set if code is illformted.

**Comment by henryiii** [comment]
One of the big changes recently that prompted me to try this again is the addition of pre-commit.ci which automatically applies any pre-commit changes that are made to a PR; this makes it easier for contributors to contribute than if black was not applied, because it corrects their style issues immediately, saving review time, no more "please fix these style issues" requests. It's been fantastic on the other projects I work on that use black.

File: bin/bump_version.py
Comment (Czaki) [reply bin/bump_version.py:62]: documentation tolds tkat black respect yapfdirectives https://black.readthedocs.io/en/stable/the_black_code_style.html?highlight=fmt%20on%20#code-style and yapf has one line disable https://github.com/google/yapf#why-does-yapf-destroy-my-awesome-formatting with `#  yapf: disable`. You may check

**Comment by joerick** [comment]
> That's the point of Black. No choices, no arguments, one style everywhere.

Yes, I do really like this about how they designed it. 

> One of the big changes recently that prompted me to try this again is the addition of pre-commit.ci which automatically applies any pre-commit changes that are made to a PR; this makes it easier for contributors to contribute than if black was not applied, because it corrects their style issues immediately, saving review time, no more "please fix these style issues" requests. It's been fantastic on the other projects I work on that use black.

yeah, that's an awesome feature of pre-commit.ci. It really feels like pre-commit inside a Github PR workflow is a great sweet-spot - keeping code quality high without discouraging new contributors

> If it got to 80%, I'd be quite happy, but not sure there. Currently for me it's seems more like 20% good, 50% meh, 30% "yikes, really?"

Yeah, me too :( it sucks, because I really want to like it - I love the idea of auto-standardised style, and all the auxiliary benefits that come with it. But it doesn't seem worth the many style regressions that I see in this diff.

**Comment by henryiii** [comment]
Let me take a pass (not quite now, but in the near future) over a few things that I think will look better with simple changes (like the `f"--message={msg}",` vs `"-m", msg,` change), if those look fine, then maybe the diff will look better.

**Comment by joerick** [comment]
> Let me take a pass (not quite now, but in the near future) over a few things that I think will look better with simple changes (like the `f"--message={msg}",` vs `"-m", msg,` change), if those look fine, then maybe the diff will look better.

Sounds good, but please don't sink too much time into it! :)

**Comment by Czaki** [comment]
There are also others formatters. Maybe yappf or another one allows creating configuration which will be acceptable.

**Comment by YannickJadoul** [comment]
> One thing which I remember from the time where I was a more active contributor to `cibuildwheel` there was a pain of commits which fixes small style problem and push on queue another job which delays all active PR.

I think that's also gotten a lot better now we're pushing way less jobs on AppVeyor, btw ;-)

> no more "please fix these style issues" requests.

To be fair, I don't think these were ever there that often, but yeah.

File: bin/bump_version.py
Comment (henryiii) [reply bin/bump_version.py]: Black now supports single-line ignores, `# fmt: skip`.

**Review by henryiii — COMMENTED** [review]
If you like it, we can polish later - you really have quite a bit of freedom; beside the trailing comma, you can use temporarily variables and things like that to provide clearer, more readable code that is written with nice Black formatting in mind.

Keep in mind most of the good changes have already gone in, so this is buying us automatic formatting at the cost of a few extra lines. It's focused on readability and reducing diffs over saving a few lines.

If you do like it, I'll probably reenable the string formatting to give us nice uniform strings before merging. But this keeps the diff simpler.

**Review by joerick — COMMENTED** [review]
Ya know, I'm coming around. I think this could work. Good work on this @henryiii. 

I've only really had a skim so far. I'm avoiding commenting on places where the code style just looks weird to me, because I'm sure I can get used to it. 

There are a few places that I noted where it seems like an obvious regression, and maybe there's a different code structure that could improve them?

But overall, I'd be okay accepting a slight regression in style in exchange for the other benefits of auto-formatting. 

By the way, I'd not be opposed to the double-quote string normalisation either. But it's smart to do that as another step, as that diff is gonna be even more crazy.

**Comment by YannickJadoul** [comment]
> But overall, I'd be okay accepting a slight regression in style in exchange for the other benefits of auto-formatting.

My biggest worry is that it's going to be annoying for some future code, though. I just don't like a computer telling me how to format my code.

**Comment by joerick** [comment]
Do you use Black in any other projects, @YannickJadoul ? I'm curious if this worry is based on experience or not :)

File: test/test_dependency_versions.py
Comment (Czaki) [reply test/test_dependency_versions.py:60]: Maybe it could be saved by adding an additional variable. But I think that it should be reported as a bug to black maintainers.

**Comment by YannickJadoul** [comment]
> Do you use Black in any other projects, @YannickJadoul ? I'm curious if this worry is based on experience or not :)

No, that's a fair point; I don't really.
It's mostly based on me not agreeing with the thing often in these PRs, formatting code I write myself and not seeing the point of having that auto-done, and never having had any issues in cibuildwheel before on style (right, or am I forgetting stuff?).
On the other hand, maybe it would make the project seem slightly more mature to contributors? But then I'm not sure formatting would do that. I guess it's a habit, after all.

**Comment by Czaki** [comment]
@YannickJadoul I participate with projects formatted using black, and it saves much time on lack of reformating and no solving merge conflicts because of change in formatting.

**Comment by joerick** [comment]
> It's mostly based on me not agreeing with the thing often in these PRs, formatting code I write myself and not seeing the point of having that auto-done, and never having had any issues in cibuildwheel before on style (right, or am I forgetting stuff?).

I agree, we've not really had problems regarding style. Probably the most we've discussed it is in the context of black!

I've started using a little recently on [pyinstrument][], and one benefit I've found is in the workflow - it is quite nice to write code freely, then invoke the editor shortcut to format the whole file.

I must say that personally, I'm more of a 'craftsperson' when it comes to writing code - auto-reformatting isn't really my style. If I was the only person working on cibuildwheel, it's unlikely that I'd be interested in it. But cibuildwheel is more of a community-driven project, and it feels like the community is gravitating towards Black for formatting, so if that is gonna help people read and contribute to the code going forward, all the better.

[pyinstrument]: /joerick/pyinstrument

File: test/test_dependency_versions.py
Comment (henryiii) [reply test/test_dependency_versions.py]: This is a use-once variable, so I just shortened the name a bit to `is_macos_11_or_later`. No need to overkill!

File: test/test_cpp_standards.py
Comment (henryiii) [reply test/test_cpp_standards.py]: Yes, and that removes the ugly "not even on the same screen" parenthesis matching too. :)

File: test/test_macos_archs.py
Comment (henryiii) [reply test/test_macos_archs.py]: Sure, use the proper Python 3 syntax instead of this weird construction. `{*a, *b}` - as a side benefit, it can now be a set! (Probably this should return a set in the first place...)

File: test/test_manylinuxXXXX_only.py
Comment (henryiii) [reply test/test_manylinuxXXXX_only.py]: It doesn't really work here anyway, since it's no more or less relevant to this line than the PyPy comment above. Same line comments should be avoided most of the time.

**Comment by henryiii** [comment]
I find future code to be better with Black, since you write it knowing that it will be Blacked. And you can always opt out if you have to. I work on lots of Blacked code, such as most of the 30 or so Scikit-HEP projects, Plumbum, and Build. Pip uses it, etc.

Benefits:
- Fast workflow: no need to prefect list indentation, **no fighting flake8 over the number of blank lines**
- Easy for contributors: pre-commit will beautify automatically
- Standard: easy to read, uniform across projects
- Encourages good practices: hard to format code is often bad code that should be refactored anyway
- Simpler diffs: The style is designed explicitly to reduce diffs
- Reduced conflicts: Two people making the same change are more likely to have the same diff

Drawbacks:
- Not as perfect as carefully crafted style. But do you ever have perfectly crafted style? Uniformly? Can you maintain it? Is it worth it? Even in your test suite?
- A little weird with trailing comments
- Wrapping can be a little weird

You can still "craft" and make the code your own by _changing_ the code, not the whitespace (trailing comma, temp variables, etc)

**Comment by henryiii** [comment]
> It's mostly based on me not agreeing with the thing often in these PRs, formatting code I write myself and not seeing the point of having that auto-done, and never having had any issues in cibuildwheel before on style (right, or am I forgetting stuff?).

This is pretty much the only repo where I have to spend several minutes after writing something fighting with flake8 adding/removing blank lines and such. Yes, our style is nice, but that's because we force things with flake8. Auto-formatting saves a ton of time.

**Comment by joerick** [comment]
I think let's just pull the plaster and get this merged. Then we can move on and get on with proper stuff like the v1.x branch, merging #596 , #438 and preparing for 2.0.

I don't have time to go over this with a fine-tooth comb (and it's probably not worth it, generally, it looks fine). The one thing I think we should do before merge is fix the misaligned triple-quote blocks. I can help with that @henryiii if you want?

**Comment by mayeut** [comment]
> I don't have time to go over this with a fine-tooth comb (and it's probably not worth it, generally, it looks fine). The one thing I think we should do before merge is fix the misaligned triple-quote blocks. I can help with that @henryiii if you want?

I had time to look at this and tried to fix those. Feel free to remove the commit (or comment) if It does not fit what you were asking for.

**Comment by mayeut** [comment]
@henryiii, I had to force push to your branch in order to resolve a merge conflict by rebasing on master.

File: bin/update_dependencies.py
Comment (joerick) [inline bin/update_dependencies.py:53]:
```suggestion
                "--volume={os.getcwd()}:/volume",
```

I don't think the quotes are required - there is no shell to interpret them, and it's not a problem if there are spaces in this path since it'll come into the child process as a single argument anyway. I discovered this in aa12480ff0e5381eca2258a6957aea6af5c46172, note the quotes on the commit message. p.s. already fixed a couple of these in https://github.com/joerick/cibuildwheel/commit/571629607169b4fa5d89304fdda1013ebf310fcd.

**Review by joerick — APPROVED** [review]
Aside from the one change, I think this is good-to-go.

**Comment by joerick** [comment]
By the way, I just learned about 'two-dot' comparisons on Github, e.g. https://github.com/joerick/cibuildwheel/compare/naive-black..henryiii:style/black - they let you compare two commits without the auto-ancestor stuff that Github does in PRs. You have to edit the URL to get them!

**Comment by henryiii** [comment]
Applied the last suggestion to my first commit, and added a final commit that activates string normalization. Ready to rebase and merge if all is well.

**Pull request merged**
